### PR TITLE
Add support for OpenBSD, and also add some regression tests.

### PR DESCRIPTION
**Pull request by buzzdeee**
Add info about OpenBSD support to README.md and metadata.json

Thanks to the lots of feedback from the other pull requests, here is a new try, following the suggestsions. 

I first came up with something like:

Facter.add(:java_version) do
...
something for non-OpenBSD
...
end
Facter.add(:java_version) do
confine :osfamily => 'OpenBSD'
...
end

that was easy, worked well on a machine that had Java installed, and also spec tests were easily written and all fine. However, on a OpenBSD machine without java installed I got warnings running facter like this:

facter -p java_version                                             
Could not retrieve fact='java_version', resolution='<anonymous>': undefined method `lines' for nil:NilClass

Those are not really an option. BTW, running Puppet 3.7.4, Facter 2.3.0. Facter only just got updated in the Ports tree to 2.4.1 Today, so haven't tested with that version yet.

If installing jdk or jre, doesn't matter, the java binary I point to, is there with both variants.

Since there is no OpenBSD yet with java 1.8, I haven't bothered yet, to fiddle with confines and versions to distinguish. Since OpenBSD is not primary target for any java development, it also might still take some time for it to appear as a port.

### CODE REVIEW

**Comment by tphoney** [comment]
Thanks for the effort you have put into this pull request, however it looks like the hardcoded path for java in (lib/facter/java_version.rb) seems wrong. What if java 1.6 is installed. Is this a special case for 1.7 on openbsd that needs documented. @mhaskel found https://ask.puppetlabs.com/question/5299/how-does-one-set-the-path-used-by-facterutilresolutionexec/

`Facter::Util::Resolution.with_env("PATH" => "/opt/foo/bin:/opt/bar/bin") do`

which would allow you to specify additional paths to be set for the fact execution. Are there are any symlinks to like /usr/local/jre/bin or something that would work with whatever version of java is installed?

Thanks

**Comment by buzzdeee** [comment]
It's that there is no official java packages that work "out of the box" on OpenBSD, so people are usually installing either from packages, or building their own from ports. As ports/package, currently only 1.7, heavily patched to build/work at all, is avaialable. Therefore it's highly unlikely, someone would install a different version somehow. 

OpenBSD doesn't have such a crude /etc/alternatives or however called, as exists on a bunch of Linux distributions. Therefore, there are no links installed with the pacakge, pointing to the right version.
Either the Admin adds the directory to the PATH globally, or the user does that on his own.

Similar is true for ruby, python etc. There are multiple versions installable in parallel, but the binaries are for example called python2.7 or ruby21, so breaking assumptions that #!/usr/bin/env python or #!/usr/bin/env ruby shebangs work out of the box ;)
Also here, the admin/user can create symlinks to the preferred version.

Ports infrastructure usually takes care to fix those to point to the right stuff, but everything that doesn't get installed via port/package, must be touched manually.

At least with respect to java, the more or less same problem I have with the java_ks puppet module, to find the "keytool" binary in the same directory where the java binary is located.

To make that long story short: maybe it's really easier to dismiss this pull request at all, and I see that I get the PATH environment variable updated just before java related puppet modules come into play.
Anyways, while digging into it, I learned a bunch of new things. Just because of that, it was already 
worth ii from my side, hope didn't wasted your time too much ;)
So if you don't insist on getting it fixed within the module, feel free to just close this one. ;)

thanks,
Sebastian

File: lib/facter/java_version.rb [hunk 1]
```diff
@@ -12,8 +12,14 @@
 # Notes:
 #   None
 Facter.add(:java_version) do
-  setcode do
-    if Facter::Util::Resolution.which('java')
+  if Facter::Util::Resolution.which('/usr/local/jdk-1.7.0/jre/bin/java')
```
Comment (elyscape) [inline lib/facter/java_version.rb]: This should be a separate fact, like so:
```ruby
Facter.add(:java_version) do
  if Facter::Util::Resolution.which('java')
    setcode do
      Facter::Util::Resolution.exec('java -Xmx8m -version 2>&1').lines.first.split(/"/)[1].strip
    end
  end
end
Facter.add(:java_version) do
  confine :osfamily => 'OpenBSD'
  if Facter::Util::Resolution.which('/usr/local/jdk-1.7.0/jre/bin/java')
    setcode do
      Facter::Util::Resolution.exec('/usr/local/jdk-1.7.0/jre/bin/java -Xmx8m -version 2>&1').lines.first.split(/"/)[1].strip
    end
  end
end
```

File: spec/unit/facter/java_version_spec.rb [hunk 2]
```diff
@@ -6,22 +6,48 @@
...
+        EOS1
+        Facter::Util::Resolution.expects(:which).with("/usr/local/jdk-1.7.0/jre/bin/java").returns(true)
+        Facter::Util::Resolution.expects(:exec).with("/usr/local/jdk-1.7.0/jre/bin/java -Xmx8m -version 2>&1").returns(java_version_output)
+        Facter.value(:java_version).should == "1.7.0_71"
+      end
+    end
     context 'returns java version when java present' do
       it do
-        java_version_output = <<-EOS
+        java_version_output = <<-EOS2
```
Comment (elyscape) [inline spec/unit/facter/java_version_spec.rb]: You don't need to have different heredoc codes so long as they don't overlap, so this change is unnecessary.

**Comment by tphoney** [comment]
I think it is still a valid pull request, if you were to explain the reasoning of PATH problems in the readme. And follow what @elyscape says about separating the fact out. 
It would be good to merge.

**Comment by buzzdeee** [comment]
alright ;)
I'll see to find some time over the weekend

**Comment by elyscape** [comment]
I just noticed that it won't detect Java if you only have the JRE installed. I'll work out a solution as soon as I have access to a keyboard.

**Comment by elyscape** [comment]
This should work for the OpenBSD fact for both the JDK and the JRE:
```ruby
Facter.add(:java_version) do
  confine :osfamily => 'OpenBSD'
  Facter::Util::Resolution.with_env("PATH" => '/usr/local/jdk-1.7.0/jre/bin:/usr/local/jre-1.7.0/bin') do
    if Facter::Util::Resolution.which('java')
      setcode do
        Facter::Util::Resolution.exec('java -Xmx8m -version 2>&1').lines.first.split(/"/)[1].strip
      end
    end
  end
end
```

**Comment by buzzdeee** [comment]
Thanks @elyscape for the pointer. 
I had to change it a little bit. it seems the which('java') resolution, picks up the path, but the exec resolution doesn't for that reason, I had to adapt it a little bit.
That was easy, but getting the regression tests right, caused me some grey hair ;)
As you see, the one with the old FACTER_GEM_VERSION="~> 1.6.0" still fails. As you can see in the test, I tried to work around the differences with the case statement, but that did not helped.
I'd be happy about any hint, how that test could be made succeed too.

Tested on a box with JDK, another one with JRE, and one without any Java installed. All of of them run a, more or less current, OpenBSD 5.7-beta.

Further under Limitation in the README.markdown, I added a little note about java_version fact and OpenBSD.

File: README.markdown [hunk 3]
```diff
@@ -91,10 +91,17 @@ OpenJDK is supported on:
...
 * SLES 11 SP1, 12 
+* OpenBSD 5.6, 5.7
 
 Sun Java is supported on:
 * Debian 6
 
+### A note to OpenBSD
+OpenBSD packages install Java JRE/JDK in a unique directory structure, not linking
+the binaries to a standard directory. Because of that, the path to this location
+is hardcoded in the java_version.rb Facter fact. Whenever a Java upgrade to a newer
```
Comment (elyscape) [inline README.markdown]: This should just read:

   is hardcoded in the `java_version` fact.

File: lib/facter/java_version.rb [hunk 4]
```diff
@@ -12,9 +12,21 @@
 # Notes:
 #   None
 Facter.add(:java_version) do
-  setcode do
-    if Facter::Util::Resolution.which('java')
+  if Facter::Util::Resolution.which('java')
```
Comment (elyscape) [inline lib/facter/java_version.rb]: The order here was correct before. Why did you change it?

**Comment by elyscape** [comment]
The tests need some work. I'll take a deeper look in a little while.

File: lib/facter/java_version.rb (hunk 4 above)
Comment (buzzdeee) [inline lib/facter/java_version.rb]: Good catchk, I guess it's just a cut'n pasto error I didn't noticed before. Being busy at work today, have to check later today when back home, hope to find a bit of time then.

**Comment by tphoney** [comment]
Thanks for the effort you are putting in. Appreciated :)

**Comment by buzzdeee** [comment]
@tphoney I thank you both for your patience with me ;)

Reworked the java_version fact as suggestted, and finally the tests are now all green.
Wit the Facter::Util::Resolution.with_env within the setcode, now the saving of the result in the extra variable I had to do before, is not needed anymore. the Facter::Util::Resolution.exec now picks up the PATH too.

The confine was changed from osfamily to operatingsystem, as it seems that the first test with the old Facter 1.6.0 doesn't support the osfamily test, but it does support operatingsystem.

That is also the reason, why the tests are now all green after restructuring. However, with the first test, with FACTER_GEM_VERSION=1.6.0, I still get some errors/warnings spit out, despite they are passing. The facter 1.6 doesn't seem to support Facter::Util::Resolution.with_env.

Let me know if there is still something that requires changes/enhancements.

cheers

**Comment by buzzdeee** [comment]
just rebased...

**Comment by elyscape** [comment]
The tests still need some work.  I should have time over the next couple of days to take a look.

**Comment by jhoblitt** [comment]
Would it make sense to add a new fact for the path to the java executable?  That fact could be reused in other facts instead of having to potentially do a resolution inside multiple facts for the same value.

**Comment by tphoney** [comment]
This looks great, except for one detail i overlooked. Supported modules are only for the following OSes. (RHEL,Windows,Ubuntu,Debian,Solaris,SLES,Scientific Linux,CentOS,Oracle Linux,AIX). We can be compatible with BSD and that is ok. Can you remove the section for support in the metadata.json file as we do not officially support. It is fine to mention it in the readme, that we are compatible.

File: spec/unit/facter/java_patch_level_spec.rb [hunk 5]
```diff
@@ -17,8 +17,8 @@
       end
     end
 
-    context "if java is installed" do
-      context 'returns java patch version extracted from java_version fact' do
+    context "if java is not installed" do
+      context 'returns nil because of non existent java_version fact' do
```
Comment (elyscape) [inline spec/unit/facter/java_patch_level_spec.rb]: Why was this change made?

**Comment by buzzdeee** [comment]
Removed mentioning of OpenBSD from the metadata.json as requested. Rebased and squashed again.

File: spec/unit/facter/java_patch_level_spec.rb (hunk 5 above)
Comment (buzzdeee) [inline spec/unit/facter/java_patch_level_spec.rb]: look at the whole file ;)
While I was looking at the spec tests, I found this seemed to me to be a cut'n pasto error, that doesn't really reflect what the test is doing, nor what it expects to get back.

I can revert that if you prefer.

cheers

Comment (elyscape) [inline spec/unit/facter/java_patch_level_spec.rb]: Ah, gotcha.

Improved wording on that second line might just be be `context 'returns nil' do`.

File: spec/unit/facter/java_version_spec.rb [hunk 6]
```diff
@@ -6,22 +6,54 @@
   }
 
   describe "java_version" do
+    context 'returns java version when java present on OpenBSD' do
```
Comment (elyscape) [inline spec/unit/facter/java_version_spec.rb]: This should probably be refactored into a sub-context.  Something like:
```ruby
context 'returns java version when java present' do
  context 'on OpenBSD' do
    before do
      Facter.fact(:operatingsystem).stubs(:value).returns("OpenBSD")
    end
    let(:facts) { {:operatingsystem => 'OpenBSD'} }
    it do
      java_version_output = <<-EOS
openjdk version "1.7.0_71"
OpenJDK Runtime Environment (build 1.7.0_71-b14)
OpenJDK 64-Bit Server VM (build 24.71-b01, mixed mode)
      EOS
      Facter::Util::Resolution.expects(:which).with("java").returns('/usr/local/jdk-1.7.0/jre/bin/java')
      Facter::Util::Resolution.expects(:exec).with("java -Xmx8m -version 2>&1").returns(java_version_output)
      Facter.value(:java_version).should == "1.7.0_71"
    end
  end
  context 'on other systems' do
    before do
      Facter.fact(:operatingsystem).stubs(:value).returns("MyOS")
    end
    let(:facts) { {:operatingsystem => 'MyOS'} }
    it do
      java_version_output = <<-EOS
java version "1.7.0_71"
Java(TM) SE Runtime Environment (build 1.7.0_71-b14)
Java HotSpot(TM) 64-Bit Server VM (build 24.71-b01, mixed mode)
      EOS
      Facter::Util::Resolution.expects(:exec).with("java -Xmx8m -version 2>&1").returns(java_version_output)
      Facter.value(:java_version).should == "1.7.0_71"
    end
  end
end
```

File: spec/unit/facter/java_version_spec.rb [hunk 7]
```diff
@@ -6,22 +6,54 @@
...
 Java HotSpot(TM) 64-Bit Server VM (build 24.71-b01, mixed mode)
         EOS
-        Facter::Util::Resolution.expects(:which).with("java").returns(true)
         Facter::Util::Resolution.expects(:exec).with("java -Xmx8m -version 2>&1").returns(java_version_output)
         Facter.value(:java_version).should == "1.7.0_71"
       end
     end
 
-    context 'returns nil when java not present' do
+    context 'returns nil when java not present on OpenBSD' do
```
Comment (elyscape) [inline spec/unit/facter/java_version_spec.rb]: This should also probably be refactored into a sub-context:
```ruby
context 'returns nil when java not present' do
  context 'on OpenBSD' do
    before do
      Facter.fact(:operatingsystem).stubs(:value).returns("OpenBSD")
    end
    let(:facts) { {:operatingsystem => 'OpenBSD'} }
    it do
      Facter::Util::Resolution.stubs(:exec)
      Facter.value(:java_version).should be_nil
    end
  end
  context 'on other systems' do
    before do
      Facter.fact(:operatingsystem).stubs(:value).returns("MyOS")
    end
    let(:facts) { {:operatingsystem => 'MyOS'} }
    it do
      Facter::Util::Resolution.expects(:which).with("java").returns(false)
      Facter.value(:java_version).should be_nil
    end
  end
end
```

**Comment by buzzdeee** [comment]
updated the specs based on suggestions
